File: CSharp/HtmlWalker/Platform/ITagApi.cs
using System.Collections.Generic;

namespace HtmlWalker.Platform
{
    public abstract partial class WalkerPlatform
    {
        /// <summary>
        ///   The interface provides the platform specific implementation for the tag
        ///   adapters.
        ///
        ///   This is essentially a bridge between the platform specific implementation and
        ///   the tag adapter interface. Generally, this interface is used to in the
        ///   platform implementation but is not used directly when making use of the HTML
        ///   walker.
        /// </summary>
        /// <author>Robert R Flores</author>
        public interface ITagApi
        {
            /// <summary>
            ///   Returns the name of the housed HTML tag, such as 'html', 'body', 'p', etc
            /// </summary>
            string Name(object tag);

            /// <summary>
            ///   Access to the walker platform object specific to the HTML Walker implementation
            /// </summary>
            WalkerPlatform Platform { get; }

            /// <summary>
            ///   Gives the platorm specific way to get a tags inner text
            /// </summary>
            string GetText(object tag);

            void SetText(object tag, string text);

            string OwnedText(object tag);

            /// <summary>
            ///   Gives the platform specific way to get the 'script' text
            ///   in such HTML tags such as <b>&lt;script&gt;</b> and <b>&lt;style&gt;</b>
            /// </summary>
            string GetScript(object tag);

            void SetScript(object tag, string script);

            /// <summary>
            ///   Gives the platfrom specifice way to get attributes, used by <see cref="BaseTag.Attributes"/>
            ///   to implement <see cref="ITagApi.GetAttributes"/>
            /// </summary>
            /// <returns>map of attributes to values</returns>
            IDictionary<string, string> GetAttributes(object tag);

            /// <summary>
            ///   Gives the platfrom specifice way to get an attribute, used by <see cref="BaseTag.Attributes"/>
            ///   to implement <see cref="ITagApi.GetAttribute(string)"/>
            /// </summary>
            /// <param name="attr">attribute name</param>
            /// <returns>attribute value</returns>
            string GetAttribute(object tag, string attr);

            /// <summary>
            ///   Gives the platfrom specifice way to set an attribute, used by <see cref="BaseTag.Attributes"/>
            ///   to implement <see cref="ITagApi.SetAttribute(string, string)"/>
            /// </summary>
            /// <param name="attr">attribute name</param>
            /// <param name="value">attribute value</param>
            void SetAttribute(object tag, string attr, string value);

            /// <summary>
            ///   Provides the platform specific way to add an HTML tag to the content
            /// </summary>
            /// <param name="tag">the HTML tag object</param>
            void Add(object tag, ITag adapter);

            void Remove(object tag, ITag adapter);

            void Insert(object tag, int index, ITag adapter);
        }
    }
}

File: Java/htmlwalker/src/main/java/htmlwalker/platform/ITagApi.java
/*
 * 
 */
package htmlwalker.platform;

import java.util.Map;

import htmlwalker.ITag;

/**
 * The interface provides the platform specific implementation for the tag
 * adapters.
 * 
 * This is essentially a bridge between the platform specific implementation and
 * the tag adapter interface. Generally, this interface is used to in the
 * platform implementation but is not used directly when making use of the HTML
 * walker.
 * 
 * @author Robert Flores
 */
public interface ITagApi
{
	/**
	 * Access to the walker platform object used to generate this instance.
	 *
	 * @return the walker platform
	 */
	public WalkerPlatform platform();

	/**
	 * Returns the tag name.
	 *
	 * @return the string
	 */
	public String name(Object tag);

	/**
	 * Provides the implementation for the <b>ITag.getAttributes()</b>
	 *
	 * @return the attributes
	 * 
	 * @see HtmlWalker.ITag#getAttributes()
	 */
	public Map<String, Object> getAttributes(Object tag);

	/**
	 * Gets the tag attribute value for a given attribute name.
	 *
	 * @param attr - the attribute name
	 * @return the attribute value
	 * 
	 * @see HtmlWalker.ITag#getAttribute(String)
	 */
	public String getAttribute(Object tag, String attr);

	/**
	 * Sets the tag attribute value for a given attribute name.
	 *
	 * @param attr - the attribute name
	 * @param value - the attribute value
	 * 
	 * @see HtmlWalker.ITag#setAttribute(String, String)
	 */
	public void setAttribute(Object tag, String attr, Object value);

	/**
	 * Gets the text for tags of content type <b>TEXT</b>.
	 *
	 * @return the text belonging to the HTML tag
	 * 
	 * @see HtmlWalker.ITag#getInnerText()
	 */
	public String getText(Object tag);

	/**
	 * Sets the text for tags of content type <b>TEXT</b>.
	 *
	 * @param text - new text
	 */
	public void setText(Object tag, String text);

	/**
	 * Gets the script for tags of content type <b>SCRIPT</b>.
	 *
	 * @return the script belonging to the HTML tag
	 * 
	 * @see HtmlWalker.ITag#getInnerText()
	 */
	public String getScript(Object tag);

	/**
	 * Sets the script for tags of content type <b>SCRIPT</b>.
	 *
	 * @param script - new script
	 */
	public void setScript(Object tag, String script);
	
	public void add(Object tag, ITag adapter);
	
	public void remove(Object tag, ITag adapter);

	public String ownedText(Object tag);
}
